x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,359 @@
+        this.secureRandom = SecureRandom.getInstance(EncryptedRepositoryPlugin.RAND_ALGO);
+        this.passwordSettingForThisRepo =
+                (SecureSetting<?>) EncryptedRepositoryPlugin.ENCRYPTION_PASSWORD_SETTING.getConcreteSettingForNamespace(metadata.name());
+    }
+
+    @Override
+    public void snapshotShard(Store store, MapperService mapperService, SnapshotId snapshotId, IndexId indexId,
+                              IndexCommit snapshotIndexCommit, IndexShardSnapshotStatus snapshotStatus, boolean writeShardGens,
+                              ActionListener<String> listener) {
+        if (EncryptedRepositoryPlugin.getLicenseState().isEncryptedRepositoryAllowed()) {
+            if (consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+                super.snapshotShard(store, mapperService, snapshotId, indexId, snapshotIndexCommit, snapshotStatus, writeShardGens, listener);
+            } else {
+                listener.onFailure(new RepositoryException(metadata.name(),
+                        "Password mismatch for repository [" + metadata.name() + "]. The local node's value of the " +
+                                "keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] is different from the master's"));
+            }
+        } else {
+            listener.onFailure(LicenseUtils.newComplianceException(
+                    EncryptedRepositoryPlugin.REPOSITORY_TYPE_NAME + " snapshot repository"));
+        }
+    }
+
+    @Override
+    public void restoreShard(Store store, SnapshotId snapshotId, IndexId indexId, ShardId snapshotShardId,
+                             RecoveryState recoveryState, ActionListener<Void> listener) {
+        if (false == consistentSettingsService.isConsistent(passwordSettingForThisRepo)) {
+            // the repository has a different password on the local node compared to the master node
+            // even though restoring the shard will surely fail (because we know that, by now, the master's password
+            // must be correct, otherwise this method will not get called) we let it pass-through in order to avoid
+            // having to manipulate the {@code recoveryState} argument
+            logger.error("Password mismatch for repository [" + metadata.name() + "]. The local node's value of the " +
+                    "keystore secure setting [" + passwordSettingForThisRepo.getKey() + "] is different from the master's");
+        }
+        super.restoreShard(store, snapshotId, indexId, snapshotShardId, recoveryState, ActionListener.delegateResponse(listener,
+                (l, e) -> l.onFailure(new RepositoryException(metadata.name(), "Password mismatch for repository [" + metadata.name() +
+                        "]. The local node's value of the keystore secure setting [" +
+                        passwordSettingForThisRepo.getKey() + "] is different from the master's"))));
+    }
```

Comment:
@original-brownbear What do you think about this, in the case where we know the restoring will fail because the data node has the wrong password? I think attempting the restore is a more "robust" way.

Here are the logs, on the node:
```
[2020-01-15T15:19:17,865][WARN ][o.e.i.c.IndicesClusterStateService] [ElasticMBP.local] [twitter2][0] marking and sending shard failed due to [failed recovery]
org.elasticsearch.indices.recovery.RecoveryFailedException: [twitter2][0]: Recovery failed on {ElasticMBP.local}{pN1b441YTV6fKug6AkF2Vg}{uHr4xFDrQzWtykhxV_VO9Q}{127.0.0.1}{127.0.0.1:9301}{dilm}{ml.machine_memory=17179869184, xpack.installed=true, ml.max_open_jobs=20}
        at org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$21(IndexShard.java:2606) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.lambda$recoveryListener$6(StoreRecovery.java:350) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.lambda$restore$8(StoreRecovery.java:469) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:491) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:278) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1850) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2553) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:688) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
        ... 14 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: restore failed
        ... 12 more
Caused by: org.elasticsearch.repositories.RepositoryException: [test_enc] could not read repository data from index blob
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1167) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1073) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:482) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        ... 9 more
Caused by: java.io.IOException: Failure to decrypt metadata for blob [index-6]
        at org.elasticsearch.repositories.encrypted.EncryptedRepository$EncryptedBlobContainer.readBlob(EncryptedRepository.java:243) ~[?:?]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1153) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1073) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:482) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        ... 9 more
Caused by: javax.crypto.AEADBadTagException: Tag mismatch!
        at com.sun.crypto.provider.GaloisCounterMode.decryptFinal(GaloisCounterMode.java:623) ~[?:?]
        at com.sun.crypto.provider.CipherCore.finalNoPadding(CipherCore.java:1116) ~[?:?]
        at com.sun.crypto.provider.CipherCore.fillOutputBuffer(CipherCore.java:1053) ~[?:?]
        at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:853) ~[?:?]
        at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:446) ~[?:?]
        at javax.crypto.Cipher.doFinal(Cipher.java:2266) ~[?:?]
        at org.elasticsearch.repositories.encrypted.PasswordBasedEncryptor.decrypt(PasswordBasedEncryptor.java:137) ~[?:?]
        at org.elasticsearch.repositories.encrypted.EncryptedRepository$EncryptedBlobContainer.readBlob(EncryptedRepository.java:237) ~[?:?]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1153) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1073) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:482) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        ... 9 more
[2020-01-15T15:19:17,870][WARN ][o.e.i.c.IndicesClusterStateService] [ElasticMBP.local] [twitter][0] marking and sending shard failed due to [failed recovery]
org.elasticsearch.indices.recovery.RecoveryFailedException: [twitter][0]: Recovery failed on {ElasticMBP.local}{pN1b441YTV6fKug6AkF2Vg}{uHr4xFDrQzWtykhxV_VO9Q}{127.0.0.1}{127.0.0.1:9301}{dilm}{ml.machine_memory=17179869184, xpack.installed=true, ml.max_open_jobs=20}
        at org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$21(IndexShard.java:2606) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.lambda$recoveryListener$6(StoreRecovery.java:350) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.lambda$restore$8(StoreRecovery.java:469) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:71) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:491) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:278) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1850) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2553) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:73) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:688) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
        ... 14 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: restore failed
        ... 12 more
Caused by: org.elasticsearch.repositories.RepositoryException: [test_enc] could not read repository data from index blob
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1167) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1073) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:482) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        ... 9 more
Caused by: java.io.IOException: Failure to decrypt metadata for blob [index-6]
        at org.elasticsearch.repositories.encrypted.EncryptedRepository$EncryptedBlobContainer.readBlob(EncryptedRepository.java:243) ~[?:?]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1153) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1073) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:482) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        ... 9 more
Caused by: javax.crypto.AEADBadTagException: Tag mismatch!
        at com.sun.crypto.provider.GaloisCounterMode.decryptFinal(GaloisCounterMode.java:623) ~[?:?]
        at com.sun.crypto.provider.CipherCore.finalNoPadding(CipherCore.java:1116) ~[?:?]
        at com.sun.crypto.provider.CipherCore.fillOutputBuffer(CipherCore.java:1053) ~[?:?]
        at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:853) ~[?:?]
        at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:446) ~[?:?]
        at javax.crypto.Cipher.doFinal(Cipher.java:2266) ~[?:?]
        at org.elasticsearch.repositories.encrypted.PasswordBasedEncryptor.decrypt(PasswordBasedEncryptor.java:137) ~[?:?]
        at org.elasticsearch.repositories.encrypted.EncryptedRepository$EncryptedBlobContainer.readBlob(EncryptedRepository.java:237) ~[?:?]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1153) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getRepositoryData(BlobStoreRepository.java:1073) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:482) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
        ... 9 more
```
But the restore response is confusing:
```
curl -u elastic:password -X POST "localhost:9200/_snapshot/test_enc/snapshot_3/_restore?pretty&wait_for_completion=true"
{
  "snapshot" : {
    "snapshot" : "snapshot_3",
    "indices" : [
      "twitter",
      "twitter2"
    ],
    "shards" : {
      "total" : 3,
      "failed" : 0,
      "successful" : 3
    }
  }
}
```
I think the confusion is true regardless of the reason restoring the shard?  I feel there should be way for the caller to learn that restoring the shard failed, but I don't know why it doesn't work in this case. Can you help me please?

Reply:
The fact that we get no failed counts in the response here looks like a bug. Almost looks like something is broken with the way the exceptions are are wrapped and unwrapped by `org.elasticsearch.index.shard.StoreRecovery#recoveryListener`. I will look into this as soon as I can, but `org.elasticsearch.index.shard.StoreRecovery#recoveryListener` is probably a good start to see how that exception is dealt with.

Reply:
Thank you Armin! I will also look delve deeper given your confirmation.

Reply:
@original-brownbear Debugging further I've learned that restoring to a node with a different password (and which cannot decrypt the blobs) is an allocation failure, but it's not "fatal" as if the snapshot were two versions behind. Instead the allocation is retried (https://github.com/elastic/elasticsearch/blob/b02119c42377b191cc97808c814cd4399da99dd0/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L625). And then it's retried on a node that has the correct password, were it succeeds. Afterwards I assume the shard can be rebalanced around. Does this scenario sound plausible to you?

At first glance this might look "nice"-ish, because apparently we can tolerate some nodes with wrong passwords during restore, but there is no allocation decider to say where shards from a certain repository can be restored to; It's a game of try and miss. And the decider sounds too complicated to implement.

Instead, I have a better idea! I see `Repository#getRepositoryData` is always called on the master node before the snapshot and restore operations (and others) fan-out. I should be adding the license and password consistency check there, and hence fail the operation ASAP before it even gets to the restore and snapshot repository callbacks on the data nodes. (I need to make the password consistency check work on the master node) Does this sound correct to you?

Reply:
> Instead, I have a better idea! I see Repository#getRepositoryData is always called on the master node before the snapshot and restore operations (and others) fan-out. I should be adding the license and password consistency check there, and hence fail the operation ASAP before it even gets to the restore and snapshot repository callbacks on the data nodes. (I need to make the password consistency check work on the master node) Does this sound correct to you?

Yea I like that idea, that will also generally make the repository unusable until the setting consistency is fixed which is a good thing IMO. It would also make snapshot creation fail fast btw so that problem would be dealt with automatically also :)

Reply:
Thank you Armin!

Thinking further, if we go with this idea, then an encrypted repository would be unusable (wrt all operations) if the password is not the same across the nodes.

An argument can be made that we're overdoing it, and that technically we could allow other operations, such as listing, deleting or cleaning-up snapshots, if only the password on the master node coincides with the true password of the blob store repository (disregarding the password on the data nodes). But I think users relying on this feature is a bad thing, because the master can change at any time so the availability of those operations is unreliable. Moreover those are not crucially important operations, i.e. you do not install a repository on a big cluster (onto which you have trouble to consistently set the password) just so you can list, delete or clean-up a repository; you can do that on a one node cluster on the laptop if you wish.

Another argument could be made that, if the repository is completely unusable, then we should maybe fail earlier, for example not creating the repository in the first place or failing nodes to join the cluster. But I would say that these two behaviors lead to poorer user experience. Not creating the repository after a cluster restart leads to confusion when trying to use a missing repository; the caller doesn't get a response back about why the repository is not there, only that it's missing and it has to look in the logs. Also, failing a node to join a cluster because of the repository password mismatch is to big of a heavy-handed approach; the node can work just fine as part of the cluster, maybe it's hastily provided by some process that doesn't have access to the secrets in the keystore, it's only a non-crucial feature which is unusable.

Therefore, I would continue to implement this plan, of disabling all operations on the repository until passwords are consistent, unless @tvernum or @jkakavas have anything against it. This requires making the password consistency check work on the master node. I would also hope that in this case we could make the repository plugin reloadable so that we can update the password without restarting the node, but I'm not yet sure about it yet depends how things fall into place.

Reply:
> Therefore, I would continue to implement this plan, of disabling all operations on the repository until passwords are consistent, unless @tvernum or @jkakavas have anything against it

I am not opposed. We can always come back and relax that requirement in the future if we find a compelling reason. But if we don't do it from the beginning, we will find that some people don't have consistent passwords for some reason (e.g. allocation rules mean that the shards they want to encrypt aren't on those nodes) and then it will be harder to enforce it.

Reply:
I concur. I like this as a strict default and Tim's argumentation line on the possibility of relaxing ( as opposed to the possibility of restricting ) is spot on

Reply:
In the light of https://github.com/elastic/elasticsearch/pull/50846#discussion_r369716572 we would lose the ability to check that the password is the same across the data nodes doing the shards restore, before the restore process is started.
In this case, some nodes are not able to restore shards from the given repository. From my debugging and reading of the code, I believe a failed restore on a node will be allocated to another node, without any "smartiness" involved (i.e. it could be allocated to another node which also has the wrong password). This again is **the same problem as when cloud-based repositories have wrong client credentials**.

Therefore, the code to better handle this case should not be specific to the encrypted repository. I would suggest we consider this a different issue and handle separately.